### PR TITLE
HPCC-22186 Enable retain superfile structure for superindexes

### PR DESCRIPTION
**Pull request by kunalaswani**
Copy Added to superfile details, along with retain structure set to true

Signed-off-by: Kunal Aswani <Kunal.Aswani@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [x] I have tested my changes in multiple modern browsers
  - [x] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by kunalaswani** [comment]
@miguelvazq and @GordonSmith please review, I have added copy functionality to the superfiledetailswidget, we default check the retainsuperfilecheckbox to true, and make it readonly if the subfiles are greater than 1, and I've added a message when users hover as to why this is as per @wangkx suggestion. 

If criteria are not met, then the readonly is removed and the user can set checked to false

File: esp/src/eclwatch/SFDetailsWidget.js [hunk 1]
```diff
@@ -106,14 +110,19 @@ define([
...
+                    request: domForm.toObject(this.id + "CopyForm")
+                }).then(function (response) {
+                    context._handleResponse("CopyResponse.result", response);
+                });
+                registry.byId(this.id + "CopyDropDown").closeDropDown();
+            }
         },
+        // _onCopyCancel: function (event) {
+        //     registry.byId(this.id + "CopyDropDown").closeDropDown();
+        // },
```
Comment (GordonSmith) [inline esp/src/eclwatch/SFDetailsWidget.js]: Remove commented code (is CopyCancel needed?)

File: esp/src/eclwatch/SFDetailsWidget.js [hunk 2]
```diff
@@ -151,6 +164,11 @@ define([
                 this.logicalFile.refresh();
             }
             this.subfilesGrid.startup();
+
+            if (this.logicalFile.isSuperfile === true && this.logicalFile.NumOfSubfiles > 1 && this.logicalFile.KeyType != "null") {
```
Comment (GordonSmith) [inline esp/src/eclwatch/SFDetailsWidget.js]:
```suggestion
            if (this.logicalFile.isSuperfile === true && this.logicalFile.NumOfSubfiles > 1 && this.logicalFile.KeyType !== "null") {
```

**Review by GordonSmith — CHANGES_REQUESTED** [review]
Minor changes.

File: esp/src/eclwatch/SFDetailsWidget.js (hunk 2 above)
Comment (GordonSmith) [reply esp/src/eclwatch/SFDetailsWidget.js]: Us "null" an esp response thing?

File: esp/src/eclwatch/SFDetailsWidget.js (hunk 1 above)
Comment (kunalaswani) [reply esp/src/eclwatch/SFDetailsWidget.js]: no, it's not, was meant to be removed.

File: esp/src/eclwatch/SFDetailsWidget.js (hunk 2 above)
Comment (kunalaswani) [reply esp/src/eclwatch/SFDetailsWidget.js]: null is returned if there is no value, however could use "", i dont believe it is esp specific, will test and fix

**Comment by kunalaswani** [comment]
@GordonSmith changes made and tested

File: esp/src/eclwatch/SFDetailsWidget.js (hunk 2 above)
Comment (GordonSmith) [reply esp/src/eclwatch/SFDetailsWidget.js]: Big difference between:
```
this.logicalFile.KeyType != "null"
```
and
```
this.logicalFile.KeyType !== ""
```
?

Comment (kunalaswani) [reply esp/src/eclwatch/SFDetailsWidget.js]: so we expect 3 different key type values, if not returned, we get an undefined or "null" value, is there a better way to handle?

Comment (GordonSmith) [reply esp/src/eclwatch/SFDetailsWidget.js]: So the possible KeyType values can be:
* undefined
* "null"
* "SomeOtherKeyType"
?

In which case wouldn't your logic need to be:
```
(this.logicalFile.isSuperfile === true && this.logicalFile.NumOfSubfiles > 1 && this.logicalFile.KeyType !== "null"&&this.logicalFile.KeyType !== undefined)
```
?
(assuming I understood correctly)

Comment (kunalaswani) [reply esp/src/eclwatch/SFDetailsWidget.js]: should just be undefined, think i was just confusing myself on that one.

**Review by GordonSmith — APPROVED** [review]
@richardkchapman looks good.

**Pull request merged**